File: python/warmup-1/front3.py
# Файл: python/warmup-1/front3.py

# --- Условие ---
# Given a string, we'll say that the front is the first 3 chars of the
# string. If the string length is less than 3, the front is whatever
# is there. Return a new string which is 3 copies of the front.
#
# front3('Java') → 'JavJavJav'

# --- Мой анализ ---
# 1. Понимание: Взять первые 3 символа (или меньше, если строка короче)
#    и повторить их трижды.
# 2. Крайние случаи: Строки длиной 2, 1, 0. Срез `[:3]` в Python
#    элегатно обрабатывает все эти случаи без ошибок.
# 3. Сложность:
#    - Временная: O(1), т.к. работаем с подстроками константной длины.
#    - Пространственная: O(1), т.к. результат имеет константную макс. длину.
# 4. Альтернативы: Можно было бы написать if/else для проверки длины,
#    но это избыточно и не "по-питоновски".

# --- Решение ---


def front3(str):
    front = str[:3]
    return front * 3

File: java/warmup-1/Front3.java
// Файл: java/warmup-1/Front3.java

public class Front3 {

    /*
     --- Условие ---
     Given a string, we'll say that the front is the first 3 chars of the
     string. If the string length is less than 3, the front is whatever
     is there. Return a new string which is 3 copies of the front.

     --- Мой анализ ---
     1. Понимание: Логика та же, что и в Python, но реализация требует
        явной проверки длины строки, чтобы избежать ошибки IndexOutOfBoundsException.
     2. Крайние случаи: Строки длиной < 3. Обрабатываются через if.
     3. Сложность:
        - Временная: O(1).
        - Пространственная: O(1).
     4. Альтернативы: В Java 11+ появился метод `String.repeat(n)`, который
        позволил бы написать `return front.repeat(3);`. Конкатенация
        `front + front + front` работает во всех версиях Java.
    */

    // --- Решение ---
    public String front3(String str) {
        String front;

        if (str.length() < 3) {
            front = str;
        } else {
            front = str.substring(0, 3);
        }

        return front + front + front;
    }

    // --- Метод для самопроверки ---
    public static void main(String[] args) {
        Front3 solution = new Front3();

        System.out.println("Тестирование метода front3:");
        // Ожидаем: JavJavJav
        System.out.println(solution.front3("Java"));
        // Ожидаем: ababab
        System.out.println(solution.front3("ab"));
        // Ожидаем: aaa
        System.out.println(solution.front3("a"));
    }
}
